### PR TITLE
Doc tuning parameters

### LINKED ISSUE #80: Analyze the results of hyper parameter tuning stt
For SST variable, we performed Hyper parameter tuning using Ray tune library with these [strategy](https://github.com/ESMValGroup/ClimaNet/issues/40#issuecomment-4877040064). The config ray is in its [script](https://github.com/ESMValGroup/ClimaNet/blob/main/scripts/tuning.py) and the results are stored on Levante in this path `/work/<account_id>/eso4clima/tune/sst_01`.  After merging #78, we want to analyze the results as:

- run the best model on test data and document loss
- document the hyper parameter of the best model --> those parameters should be used for training in #25

### PR DESCRIPTION
**Pull request by rogerkuou**
fix #80 

Added a README file documenting the tuning parameters and testing results on the best tuned model.

Tuning testing results are in `/work/<account_id>/eso4clima/tune/run_best_model`

Testing loss on year 2022 is 0.0367

### CODE REVIEW

**Comment by rogerkuou** [comment]
Hi  @SarahAlidoost  the best tuning model has ben run against the test dataset. The loss looks okay.

Can you review this? Thanks!

File: scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Scripts
+
+## Structure
+
+- `data_preparation.*`: Scripts for preparing the data for training, tuning and evaluation. Mainly converting large netCDF files to Zarr storage with specific chunking strategies. This allows executing training for larger-than-memory datasets.
```
Comment (SarahAlidoost) [inline scripts/README.md]:
```suggestion
- `data_preparation.*`: Scripts for preparing the data for training, tuning and evaluation and saving them to Zarr storage with specific chunking strategies.
```

File: scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+- `tuning.*`: Scripts for hyperparameter tuning.
+- `run_best_tuned_model.*`: Scripts for running the best tuned model on the test set.
+
+## Experiments
+
+### Tuning experiments
+
+- datasplit: train set = 2020, validation set = 2021, test set = 2022
+- path of tuning results: `/work/<account_id>/eso4clima/tune/`.
```
Comment (SarahAlidoost) [inline scripts/README.md]:
```suggestion
- path of tuning results: `/work/<account_id>/eso4clima/tune/sst_01`.
```

File: scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+## Structure
+
+- `data_preparation.*`: Scripts for preparing the data for training, tuning and evaluation. Mainly converting large netCDF files to Zarr storage with specific chunking strategies. This allows executing training for larger-than-memory datasets.
+- `tuning.*`: Scripts for hyperparameter tuning.
+- `run_best_tuned_model.*`: Scripts for running the best tuned model on the test set.
+
+## Experiments
+
+### Tuning experiments
```
Comment (SarahAlidoost) [inline scripts/README.md]:
```suggestion
### Tuning experiments for SST variable
```

File: scripts/run_best_tuned_model.slurm [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+#SBATCH --job-name=climanet_eval
+#SBATCH --nodes=1
+#SBATCH --ntasks-per-node=1
+#SBATCH --cpus-per-task=128
+#SBATCH --time=02:00:00
+#SBATCH --account=bd0854
+#SBATCH --partition=compute
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.slurm]: This should be a gpu node.

File: scripts/run_best_tuned_model.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    test_data_zarr_dir.mkdir(parents=True, exist_ok=True)
+
+    _ = data_preparation(
+        daily_data_test[args.var_name],
+        monthly_data_test[args.var_name],
+        calculate_residuals=True,
+        is_hourly=True,
+        save_to_zarr=True,
+        run_dir=test_data_zarr_dir,
+    )
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]: since data is already prepared and stored in zarr, this part should be removed.

File: scripts/run_best_tuned_model.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    if not experiment_path.exists():
+        raise FileNotFoundError(
+            f"Experiment directory does not exist: {experiment_path}"
+        )
+    if not test_data_dir.exists():
+        raise FileNotFoundError(f"Test data directory does not exist: {test_data_dir}")
+    if not lsm_file_path.exists():
+        raise FileNotFoundError(f"LSM file does not exist: {lsm_file_path}")
+
+    daily_files = list(
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
   input_files = list(
```

Comment (SarahAlidoost) [reply scripts/run_best_tuned_model.py]: please replace daily/hourly names with `input`.

File: scripts/run_best_tuned_model.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        read_st_data(
+            data_path=test_data_zarr_dir,
+            var_name=args.var_name,
+        )
+    )
+
+    lsm_mask = xr.open_dataset(lsm_file_path)
+
+    num_patches = (10, 10)
+    patch_size = (1, 4, 4)
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]: ca you use the same  `"patch_size": (1, 40, 40)` as in [hyper parameter settings](https://github.com/ESMValGroup/ClimaNet/blob/1e93be51c12b68dba4563b70ed5f186605c29d89/scripts/tuning.py#L64).

File: scripts/run_best_tuned_model.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    spatial_patch_size = (
+        patch_size[1] * num_patches[0],
+        patch_size[2] * num_patches[1],
+    )
+    stride = (spatial_patch_size[0] // 5, spatial_patch_size[1] // 5)
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]: Can you use the same stride `"stride": (20, 20)` as in [hyper parameter settings](https://github.com/ESMValGroup/ClimaNet/blob/1e93be51c12b68dba4563b70ed5f186605c29d89/scripts/tuning.py#L65).

File: scripts/run_best_tuned_model.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    best_checkpoint = best_result.checkpoint
+    model_path = Path(best_checkpoint.path) / "checkpoint.pt"
+    print(f"Best checkpoint path: {model_path}")
+
+    prediction_config = PredictionConfig(
+        calculate_residuals=True,
+        return_numpy=True,
+        save_predictions=False,
+        return_loss=True,
+        device="cpu",
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]: please use `device="cuda"`

File: scripts/run_best_tuned_model.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+    analysis = tune.ExperimentAnalysis(str(experiment_path))
+    best_result = analysis.get_best_trial("loss", "min")
+    best_checkpoint = best_result.checkpoint
+    model_path = Path(best_checkpoint.path) / "checkpoint.pt"
+    print(f"Best checkpoint path: {model_path}")
+
+    prediction_config = PredictionConfig(
+        calculate_residuals=True,
+        return_numpy=True,
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
        return_numpy=False,
```

File: scripts/run_best_tuned_model.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        save_predictions=False,
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]: Can we save the results and compare 12 month of data?

File: scripts/run_best_tuned_model.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        verbose=False,
+    )
+
+    dataloader_config = DataLoaderConfig(
+        batch_size=10,
+        shuffle=True,
+        num_workers=0,
+        pin_memory=False,
+        persistent_workers=False,
+        device="cpu",
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]: Please adjust these settings when device is cuda.

File: scripts/run_best_tuned_model.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+    test_loss = predict_monthly_var(
+        model=model_path,
+        dataset=dataset_test,
+        dataloader_config=dataloader_config,
+        prediction_config=prediction_config,
+        run_dir=run_dir,
+    )
+
+    print("Test loss:")
+    print(test_loss)
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
    print("Test loss:", test_loss)
```

File: scripts/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    {'patch_size': 8,
+    'overlap': 1,
+    'embed_dim': 64,
+    'dropout': 0.2,
+    'hidden': 32,
+    'spatial_depth': 3,
+    'spatial_heads': 2,
+    'optimizer_lr': 0.001787422899066508,
+    'batch_config': {'batch_size': 100, 'accumulation_steps': 2}}
+ ```
```
Comment (SarahAlidoost) [inline scripts/README.md:27]: Can we add the code of extracting the parameters to the `run_best_tuned_model.py?

File: scripts/run_best_tuned_model.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+        sh_embed_dim=96,
+        sh_order_L=10,
+        verbose=True,
+        load_lazy=False,
+    )
+    print(f"Created test dataset with {len(dataset_test)} patches.")
+
+    analysis = tune.ExperimentAnalysis(str(experiment_path))
+    best_result = analysis.get_best_trial("loss", "min")
+    best_checkpoint = best_result.checkpoint
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py:134]: Can we print best parameters as well?

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@rogerkuou Thanks! Looks good 👍 I added a few comments. Let me know if something is unclear.

File: scripts/README.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+- `tuning.*`: Scripts for hyperparameter tuning.
+- `run_best_tuned_model.*`: Scripts for running the best tuned model on the test set.
+
+## Experiments
+
+### Tuning experiments
+
+- datasplit: train set = 2020, validation set = 2021, test set = 2022
+- path of tuning results: `/work/<account_id>/eso4clima/tune/`.
+- test loss: 0.036662004509047774
```
Comment (SarahAlidoost) [inline scripts/README.md]:
```suggestion
- test loss: 0.036662004509047774 (K)
```

File: scripts/README.md (hunk 14 above)
Comment (rogerkuou) [reply scripts/README.md:27]: Added in the Python scripts

File: scripts/run_best_tuned_model.py (hunk 15 above)
Comment (rogerkuou) [reply scripts/run_best_tuned_model.py:134]: Now the parameters are printed in these lines:

https://github.com/ESMValGroup/ClimaNet/blob/a843779a9ca2ac8e0bfbfc9f2d001301caa2f13d/scripts/run_best_tuned_model.py#L131-L132

File: scripts/run_best_tuned_model.py (hunk 7 above)
Comment (rogerkuou) [reply scripts/run_best_tuned_model.py]: parameters are updated:
https://github.com/ESMValGroup/ClimaNet/blob/a843779a9ca2ac8e0bfbfc9f2d001301caa2f13d/scripts/run_best_tuned_model.py#L104

File: scripts/run_best_tuned_model.py (hunk 5 above)
Comment (rogerkuou) [reply scripts/run_best_tuned_model.py]: Now this part has been removed. I updated the input arg:

https://github.com/ESMValGroup/ClimaNet/blob/a843779a9ca2ac8e0bfbfc9f2d001301caa2f13d/scripts/run_best_tuned_model.py#L25-L30

This assumes the test files are available in zarr

File: scripts/run_best_tuned_model.slurm (hunk 4 above)
Comment (rogerkuou) [reply scripts/run_best_tuned_model.slurm]: updated to use GPU partition

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost , I adapted all your suggestions. Do you want to give another look? Thanks!

File: scripts/run_best_tuned_model.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+    dataset_test = STDataset(
+        input_da=input_da,
+        input_da_nan_mask=input_da_nan_mask,
+        monthly_da=monthly_da,
+        padded_days_mask=padded_days_mask,
+        time_features=time_features,
+        land_mask=lsm_mask["lsm"],
+        patch_size=(1, *spatial_patch_size),
+        stride=stride,
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
        stride=None,
```

File: scripts/run_best_tuned_model.py [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+    lsm_mask = xr.open_dataset(lsm_file_path)
+
+    num_patches = (10, 10)
+    patch_size = (1, 40, 40)
+    spatial_patch_size = (
+        patch_size[1] * num_patches[0],
+        patch_size[2] * num_patches[1],
+    )
+    stride = (20, 20)
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
```

File: scripts/run_best_tuned_model.py (hunk 17 above)
Comment (SarahAlidoost) [reply scripts/run_best_tuned_model.py]: We don't need `stride` for inference.  This argument is mainly for creating more data samples during training.

File: scripts/run_best_tuned_model.py [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+        return_numpy=False,
+        save_predictions=False,
+        return_loss=True,
+        device="cuda",
+        verbose=False,
+    )
+
+    dataloader_config = DataLoaderConfig(
+        batch_size=10,
+        shuffle=True,
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
        shuffle=False,
```

File: scripts/run_best_tuned_model.py [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+        num_workers=0,
+        pin_memory=False,
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.py]:
```suggestion
        pin_memory=True,  # set it to True when device=cuda
```

File: scripts/run_best_tuned_model.slurm [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+#SBATCH --account=bd0854
+#SBATCH --partition=gpu
+#SBATCH --output=climanet_eval_%j.out
+#SBATCH --error=climanet_eval_%j.err
+
+set -euo pipefail
+
+source /home/b/b383704/eso4clima/ClimaNet/.venv/bin/activate
+
+python -u /home/b/b383704/eso4clima/run_best_tuned_model/run_best_tuned_model.py \
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.slurm]:
```suggestion
python -u ./Climanet/scripts/run_best_tuned_model.py \
```

File: scripts/run_best_tuned_model.slurm [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  --experiment-path /work/bd0854/eso4clima/tune/sst_01 \
+  --test-data-dir /work/bd0854/b380103/eso4clima/output/sst/concatenated/ \
```
Comment (SarahAlidoost) [inline scripts/run_best_tuned_model.slurm]: should this path point to preprocessed zarr files and not the raw data?

**Review by SarahAlidoost — APPROVED** [review]
@rogerkuou thanks. A few more comments and then it can be merged.

File: scripts/run_best_tuned_model.slurm (hunk 22 above)
Comment (rogerkuou) [reply scripts/run_best_tuned_model.slurm]: Updated to `/work/bd0854/eso4clima/preprocessed/sst/`

**Comment by rogerkuou** [comment]
Thanks @SarahAlidoost I have adapted all your comments. will merge
